### DISCUSSION TITLE
A subplot composer for plotnine.

### DISCUSSION THREAD

**Discussion by ponnhide** [has2k1/plotnine#686] [Ideas]
A subplot composer for plotnine.

Thank you for developing and managing the plotinine module.
I'm ponnhide. Maybe you know, I had independently developed patchworklib that can handle a plotnine plot as a maplotlib subplot until now. However, this time, I decided to quit support the current and future version of plotnine by patchworklib becuase it is difficult to chase the updates.

On the other hands, If you still have an plan to natively develop a subplot function in plotnine, I wanna help you using the knowledges and skills I have gained in the patchworklib development.

**Comment by has2k1** [comment]
Hello @ponnhide, patchworklib did made quite an effort in managing plotnine plots and I guess that is why it hard to keep up. But the results were great.

**Current State**

Plotnine now has a layout manager closer in concept to matplotlib's `tight_layout` than `constrained_layout`. It works well but I feel like it can be simplified some more, though I do now know how as yet.

**Next**

The goal is to make it easier to extend plotnine so that other packages can create composite panels e.g. pairplots and also have subplots (separate plots but bundled together).

**How to achieve the *Next* part?**

I am not sure yet, it is going to take some experimenting. But, we have two levels to work at

1. GridSpec
2. SubFigure

and to do so with the user API as the priority yet also mindful not to manipulate Matplotlib too deep that we later on have a hard time keeping up with changes.

**Reply by ponnhide** [reply]
In patchworklib, I try to keep using traditional functions of matplotlib without using "SubFigure" and "subplot_mosaic"  that are recently implemented.in matplotlib, because their APIs may be modified in the future.

One of the most complex aspects of implementing matplotlib is the existence of two similar operations: one for adding annotation objects such as text and legends to Figure objects (Figure level functions), and the other for adding the same elements to Axes objects (Axes level functions).

From my understanding, figure-level functions should not be used for arranging subplots because annotation objects positioned by these functions are difficult to rearrange. (An Axes object and its associated annotation objects can be rearranged using "Axes.set_position", but Figure and SubFigure objects lack such a method.)

To address this issue, in patchworklib, figure-level annotations of plotnine objects are converted into axes-level annotations. However, the conversion process is complex and not perfect, which is why I regrettably had to cease support for plotnine.

If you wish to develop a subplot-composer for plotnine similar to patchwork, cowplot, and others for ggplot2, I recommend avoiding figure-level functions and annotations. It's a bit challenging, but I believe it will enhance plotnine's flexibility for subplot purposes.

**Reply by has2k1** [reply]
You are right about figure level artists if you rely on Matplotlib's `tight_layout` or `constraint_layout` or maybe even with no layout but saving with `bbox_inches="tight"`. However, with our own layout manager, attaching artists to the figure is not an issue anymore. Note that, we do not naively place an artist onto the figure e.g. at `(0.5, 0.98)` with the hope that it does not overlap anything!

A custom layout manager became possible in [Matplotlib 3.6](https://matplotlib.org/stable/users/prev_whats_new/whats_new_3.6.0.html#new-layout-engine-module), and I think it should simplify subplot composition.

**Reply by ponnhide** [reply]
Thank you for your response.
I think I might understand the concept in plotnine v0.12 for drawing a plot.
It defines the entire figure layout just before the drawing process, so users don't need to worry about the specific positions of each subplot during their construction, correct?

I wasn't aware of such a technique, but I now understand why plotnine v0.12.x plots cannot be handled properly in patchworklib.
I've now updated patchworklib to technically handle the plotnine plots of v0.12.1 (It can't fully reproduce plotnine plots, but I believe it will be a temporary solution until a native subplot composer is available.)

I also agree that using a custom layout manager would be a good approach to developing a subplot composer for plotnine.

Thank you for your effort.

**Comment by has2k1** [comment]
By the time the layout manager is invoked, we can query the figure artists for there sizes/extents. Then we make space for them, and place them accordingly without any overlaps or the figure growing in size.
